stats/opentelemetry/client_metrics.go

```diff
@@ -0,0 +1,219 @@
+	clientMetrics.attemptRcvdTotalCompressedMessageSize = createInt64Histogram(setOfMetrics, "grpc.client.attempt.rcvd_total_compressed_message_size", meter, metric.WithUnit("By"), metric.WithDescription("Compressed message bytes received per call attempt."), metric.WithExplicitBucketBoundaries(DefaultSizeBounds...))
+	clientMetrics.callDuration = createFloat64Histogram(setOfMetrics, "grpc.client.call.duration", meter, metric.WithUnit("s"), metric.WithDescription("Time taken by gRPC to complete an RPC from application's perspective."), metric.WithExplicitBucketBoundaries(DefaultLatencyBounds...))
+
+	csh.clientMetrics = clientMetrics
+}
+
+func (csh *clientStatsHandler) unaryInterceptor(ctx context.Context, method string, req, reply any, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
+	ci := &callInfo{
+		target: csh.determineTarget(cc),
+		method: removeLeadingSlash(csh.determineMethod(method, opts...)),
+	}
+	ctx = setCallInfo(ctx, ci)
+
+	startTime := time.Now()
+	err := invoker(ctx, method, req, reply, cc, opts...)
+	csh.perCallMetrics(ctx, err, startTime, ci)
+	return err
+}
+
+// determineTarget determines the target to record attributes with. This will be
+// "other" if target filter is set and specifies, the target name as is
+// otherwise.
+func (csh *clientStatsHandler) determineTarget(cc *grpc.ClientConn) string {
+	target := cc.CanonicalTarget()
+	if f := csh.o.MetricsOptions.TargetAttributeFilter; f != nil && !f(target) {
+		target = "other"
+	}
+	return target
+}
+
+// determineMethod determines the method to record attributes with. This will be
+// "other" if StaticMethod isn't specified or if method filter is set and
+// specifies, the method name as is otherwise.
+func (csh *clientStatsHandler) determineMethod(method string, opts ...grpc.CallOption) string {
+	for _, opt := range opts {
+		if _, ok := opt.(grpc.StaticMethodCallOption); ok {
+			return method
+		}
+	}
+	return "other"
+}
+
+func (csh *clientStatsHandler) streamInterceptor(ctx context.Context, desc *grpc.StreamDesc, cc *grpc.ClientConn, method string, streamer grpc.Streamer, opts ...grpc.CallOption) (grpc.ClientStream, error) {
+	ci := &callInfo{
+		target: csh.determineTarget(cc),
+		method: removeLeadingSlash(csh.determineMethod(method, opts...)),
+	}
+	ctx = setCallInfo(ctx, ci)
+	startTime := time.Now()
+
+	callback := func(err error) {
+		csh.perCallMetrics(ctx, err, startTime, ci)
+	}
+	opts = append([]grpc.CallOption{grpc.OnFinish(callback)}, opts...)
+	return streamer(ctx, desc, cc, method, opts...)
+}
+
+func (csh *clientStatsHandler) perCallMetrics(ctx context.Context, err error, startTime time.Time, ci *callInfo) {
+	s := status.Convert(err)
+	callLatency := float64(time.Since(startTime)) / float64(time.Second)
+	if csh.clientMetrics.callDuration != nil {
```

Comment:
I remember having discussions with Yash last year with respect to this performance wise since it's on the RPC path. I think the argument is that a function call is much slower than a branch (who knows, maybe the compiler will inline)? Do you agree or disagree?

Reply:
It will be an indirect call, so an extra load and a call/return, too.  I wouldn't worry about that and optimize for the way you want the code to look.  This would not be measurable next to everything else, probably.  If you disabled all the metrics (why? wouldn't you disable the recorder instead?), I'd assume the difference in performance between the two approaches would still be negligible.

Reply:
Sounds fine behavior wise. You can check the implementation to make sure there are no locks being taken.